### PR TITLE
Enable usage of gsl::narrow with exceptions disabled

### PR DESCRIPTION
**Pull request by annagrin**
This change is based on the previous PR by @djarek:

This solution uses the approach of boost::asio to enabling usage of the
library in environments where exception usage is either prohibited
or not feasible (due to code size constraints). A function template
gsl::throw_exception has been added, which in a normal environment just
throws the exception. However, when GSL_TERMINATE_ON_CONTRACT_VIOLATION
is defined the function is only declared by gsl and the definition of
this function template must be supplied by the library's user.

Closes: #468

Signed-off-by: Damian Jarek <damian.jarek93@gmail.com>

Addition:
- made sure narrow() terminates only in GSL_TERMINATE_ON_CONTRACT_VIOLATION 
  mode, and throws in all others
- added work around for MS STL to ensure termination in no-exception mode
       - understand MSVC STL no exception macro
      -  use function static variable to set termination handler in no-exception mode (msvc)
- add compile-only tests for no-exception mode

### CODE REVIEW

File: include/gsl/gsl_assert [hunk 1]
```diff
@@ -20,6 +20,13 @@
 #include <exception>
 #include <stdexcept> // for logic_error
 
+//
+// Temporary until MSVC STL supports no-exceptions mode.
+// Currently terminate is a no-op in this mode, so we add termination behavior back
+//
+#if defined(_MSC_VER) && !_HAS_EXCEPTIONS
```
Comment (neilmacintosh) [inline include/gsl/gsl_assert]: I think you want to test if _HAS_EXCEPTIONS is `#define`d before testing its value. People catch me making that mistake all the time.

File: tests/no_exception_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <gsl/gsl_util> // for narrow_cast, at
+#include <gsl/span>     // for span, span_iterator, operator==, operator!=
+
+#include "test_noexcept.hpp"
```
Comment (neilmacintosh) [inline tests/no_exception_tests.cpp]: Everywhere else we use `.h` I think. Any reason to not follow suit here and have `test_noexcept.h`?

File: tests/test_noexcept.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "test_noexcept.hpp"
+int main()
```
Comment (neilmacintosh) [inline tests/test_noexcept.cpp]: Wouldn't you want to install a replacement to the usual "terminate" behavior in this test so you can ensure each case  ends up at the terminate-handler as expected without having to actually kill the process?

**Comment by neilmacintosh** [comment]
I really like the change overall, thanks for picking up @djarek's work here @annagrin and finishing it up. I left a couple of comments to consider but (apart from the `#if`-related one) you should feel free to merge and get to them at your leisure.

File: tests/test_noexcept.cpp (hunk 3 above)
Comment (annagrin) [reply tests/test_noexcept.cpp]: yes!

File: include/gsl/gsl_assert (hunk 1 above)
Comment (annagrin) [reply include/gsl/gsl_assert]: sure

File: tests/no_exception_tests.cpp (hunk 2 above)
Comment (annagrin) [reply tests/no_exception_tests.cpp]: copied from catch.hpp, but no reason to do it here

File: tests/test_noexcept.cpp (hunk 3 above)
Comment (annagrin) [reply tests/test_noexcept.cpp]: I am not using the catch infrastructure, need to think about the best way to implement this, so will resolve it in a later PR

**Comment by annagrin** [comment]
Addressed #if-related and .hpp comments. Will implement testing for no-exception mode separately.

**Comment by neilmacintosh** [comment]
LGTM, thanks @annagrin!

**Comment by annagrin** [comment]
this closes #468 and  #635

**Pull request merged**